R/combineWidgets.R

```diff
@@ -0,0 +1,186 @@
+#'       plot_ly(iris, x = ~Sepal.Width, type = "histogram", nbinsx = 20)
+#'     )
+#'   )
+#' }
+#'
+#' @export
+combineWidgets <- function(..., nrow = NULL, ncol = NULL, title = NULL,
+                           rowsize = 1, colsize = 1, byrow = TRUE,
+                           titleCSS = "",
+                           width = NULL, height = NULL) {
+  widgets <- lapply(list(...), function(x) {
+    if (is.null(x$preRenderHook)) return(x)
+    x$preRenderHook(x)
+  })
+  nwidgets <- length(widgets)
+
+  # Get number of rows and cols
+  if (!is.null(nrow) && !is.null(ncol) && nrow * ncol < nwidgets) {
+    stop("There are too much widgets compared to the number of rows and columns")
+  } else if (is.null(nrow) && !is.null(ncol)) {
+    nrow <- ceiling(nwidgets / ncol)
+  } else if (!is.null(nrow) && is.null(ncol)) {
+    ncol <- ceiling(nwidgets / nrow)
+  } else if (is.null(nrow) && is.null(ncol)) {
+    nrow <- ceiling(sqrt(nwidgets))
+    ncol <- ceiling(nwidgets / nrow)
+  }
+
+  ncells <- nrow * ncol
+
+  # Relative size of rows and cols
+  rowsize <- rep(rowsize, length.out=nrow)
+  colsize <- rep(colsize, length.out = ncol)
+
+  # Get the html ID of each widget
+  elementId <- sapply(widgets[1:ncells], function(x) {
+    if (is.null(x)) res <- NULL
+    else res <- x$elementId
+
+    if (is.null(res)) res <- paste0("widget", floor(runif(1, max = 1e9)))
+
+    res
+  })
+
+  # Construct the html of the combined widget
+  dirClass <- ifelse(byrow, "cw-by-row", "cw-by-col")
+
+  widgetEL <- mapply(
+    function(id, size) {
+      sprintf('<div class="cw-col" style="flex:%s;-webkit-flex:%s">
+                 <div id="%s" class="cw-widget" style="width:100%%;height:100%%"></div>
+               </div>',
+              size, size, id)
+    },
+    id = elementId,
+    size = rep(colsize, length.out = ncells)
+  )
+
+  rowsEl <- lapply(1:nrow, function(i) {
+    content <- widgetEL[((i-1) * ncol + 1):(i * ncol)]
+    sprintf('<div class="cw-row %s" style="flex:%s;-webkit-flex:%s">%s</div>',
+            dirClass, rowsize[i], rowsize[i], paste(content, collapse = ""))
+  })
+
+  content <- sprintf('<div class="cw-content %s">%s</div>',
+                     dirClass, paste(rowsEl, collapse = ""))
+
+  if(!is.null(title) && !title == "") {
+    titleEl <- sprintf('<div><h2 class="cw-title" style="%s">%s</h2></div>',
+                       titleCSS, title)
+  } else {
+    titleEl <- ""
+  }
+
+  html <- sprintf('<div class="cw-container">%s%s</div>', titleEl, content)
+
+  data <- lapply(widgets, function(x) x$x)
+  widgetType <- sapply(widgets, function(x) class(x)[1])
+
+
+  x <- list(data = data, widgetType = widgetType, elementId = elementId, html = html);
+
+  # create widget
+  combinedWidget <- htmlwidgets::createWidget(
+    name = 'combineWidgets',
+    x,
+    width = width,
+    height = height,
+    package = 'htmlwidgets'
+  )
+
+  deps <- lapply(widgets, function(x) {

```

Comment:
not huge time savings but reducing by class first and then looking up dependencies should save us some time here.  I'll try to write the code once I finish a first pass.